File: CSC310-Python/_solutions/Module03-Sorting-Lower-Bounds-and-Order-Statistics/debug02/randomized_select.py
"""Order statistics in expected linear time (CLRS Section 9.2).

randomized_select(A, p, r, i) returns the i-th smallest element of the
sub-list A[p..r] (inclusive), with i counted from 1.  It partitions
around a randomly chosen pivot and recurses into the side that must
contain the answer.  The random generator is seeded so runs repeat.
"""

import random

rng = random.Random(20)


def partition(A, p, r):
    x = A[r]
    i = p - 1
    for j in range(p, r):
        if A[j] <= x:
            i += 1
            A[i], A[j] = A[j], A[i]
    A[i + 1], A[r] = A[r], A[i + 1]
    return i + 1


def randomized_partition(A, p, r):
    k = rng.randint(p, r)
    A[k], A[r] = A[r], A[k]
    return partition(A, p, r)


def randomized_select(A, p, r, i):
    if p == r:
        return A[p]
    q = randomized_partition(A, p, r)
    k = q - p + 1                # number of elements in the low side plus pivot
    if i == k:
        return A[q]
    elif i < k:
        return randomized_select(A, p, q - 1, i)
    else:
        return randomized_select(A, q + 1, r, i - k)


def select(values, i):
    """i-th smallest of values (1-based) without disturbing the caller's list."""
    A = list(values)
    return randomized_select(A, 0, len(A) - 1, i)


def main():
    data = [12, 3, 5, 7, 4, 19, 26, 1, 8, 15]
    ranked = sorted(data)
    print("data  :", data)
    print("sorted:", ranked)
    print()
    all_ok = True
    for i in range(1, len(data) + 1):
        got = select(data, i)
        ok = got == ranked[i - 1]
        all_ok = all_ok and ok
        print(f"  {i:>2}-th smallest: {got:>3}   expected {ranked[i - 1]:>3}   {'ok' if ok else 'MISMATCH'}")
    print()
    print("all order statistics correct:", all_ok)
    n = len(data)
    print("median (i = (n+1)//2):", select(data, (n + 1) // 2))


if __name__ == "__main__":
    main()

File: CSC310-Cpp/_solutions/Module03-Sorting-Lower-Bounds-and-Order-Statistics/debug02/randomized_select.cpp
// randomized_select.cpp
// RANDOMIZED-SELECT (CLRS 9.2): finds the i-th smallest element of an array
// in expected linear time using a random pivot and Lomuto partition.
// i is 1-based: i = 1 is the minimum, i = n is the maximum.
#include <iostream>
#include <random>
#include <utility>
#include <vector>

std::mt19937 rng(7);   // fixed seed so every run is reproducible

int partition(std::vector<int>& a, int p, int r) {
    int pivot = a.at(r);
    int i = p - 1;
    for (int j = p; j < r; ++j) {
        if (a.at(j) <= pivot) {
            ++i;
            std::swap(a.at(i), a.at(j));
        }
    }
    std::swap(a.at(i + 1), a.at(r));
    return i + 1;
}

int randomizedPartition(std::vector<int>& a, int p, int r) {
    int span = r - p + 1;
    int pick = p + static_cast<int>(rng() % span);
    std::swap(a.at(pick), a.at(r));
    return partition(a, p, r);
}

// Returns the i-th smallest element of a[p..r].
int randomizedSelect(std::vector<int>& a, int p, int r, int i) {
    if (p == r) {
        return a.at(p);
    }
    int q = randomizedPartition(a, p, r);
    int k = q - p + 1;           // number of elements in the low side, including the pivot
    if (i == k) {
        return a.at(q);
    }
    if (i < k) {
        return randomizedSelect(a, p, q - 1, i);
    }
    return randomizedSelect(a, q + 1, r, i - k);
}

int main() {
    const std::vector<int> data = {29, 3, 47, 12, 8, 35, 21, 40, 15, 6, 33};
    const int n = static_cast<int>(data.size());
    // sorted copy used only to show the correct answers
    std::vector<int> sorted = data;
    for (int i = 1; i < n; ++i) {
        for (int j = i; j > 0 && sorted[j - 1] > sorted[j]; --j) {
            std::swap(sorted[j - 1], sorted[j]);
        }
    }

    std::cout << "data: ";
    for (int x : data) {
        std::cout << x << ' ';
    }
    std::cout << "\n\n";
    for (int i : {1, 3, 6, 9, 11}) {
        std::vector<int> work = data;
        int found = randomizedSelect(work, 0, n - 1, i);
        std::cout << "i = " << i << ":  select -> " << found
                  << "   (sorted[" << i << "] = " << sorted[i - 1] << ")\n";
    }
    return 0;
}
